mantidproject/mantid
# Rework Abins broadening

~~This is based on top of #26980 , do not merge until that branch has been merged~~

**Description of work.**
The existing implementation of broadening in Abins has severe problems #27043 

This work replaces the existing implementation with a new system. Several broadening-related functions are provided in a new file AbinsModules/Instruments/Broadening.py ; these are pure functions in an attempt to cut down on the "statefulness" of Abins. It remains the responsibility of an _Instrument_ to know how it should be broadened, and logic is added to dispatch automatically to an efficient broadening function.

The broadening process for these instruments is specialised and cannot use a generic convolution as the broadening kernel is _frequency-dependent_. In order to achieve acceptable performance, a new(?) approach has been developed which interpolates between a limited number of convolved spectra to approximate the results of frequency-dependent broadening. The results are illustrated in #27043 and the method is described briefly in commit messages and docstrings (with ascii art!) It may be worth preparing a technical report or paper on this, depending on how original it turns out to be.

Other methods are also made available: a null function that just resamples the data, a slow but complete summation of Gaussians and an optimised summation of truncated Gaussians which is comparable in performance to the legacy implementation. Variations are also provided that replace the Gaussian with its analytic integral to provide a normal distribution; this is more formally correct than sampling the Gaussian function as our output function is a _histogram_ but in practice the results are very similar and this could be more difficult to compare with reference calculations.

Because this PR fixes the results from Abins, many tests will be broken. Other upcoming changes will also affect results (and tests), so it feels a touch absurd to generate intermediate reference data; @Mantid-Matthew suggested that this PR could target a feature branch in order to permit a series of code reviews with known broken tests. It was decided that in this particular case it is important to get the fixes into master as soon as possible, so I will rebuild the test data and try to get it all working.

**To test:**
Unit tests and system tests have been updated.

For a clear before/after picture, run Abins on one of the sample data files. Observe that the output is now smooth, and previously was not.

If performing before/after testing it would be interesting to track performance differences; the new implementation should be faster than the old one for real-world data.

Fixes #27043 

Release notes have been added

- [x] Add independent tests for new broadening functions
- [x] Generate new test data and eyeball it for reasonableness
- [x] Wait for #26980 to be merged
- [x] Rebase on master and fix all tests
- [x] Write release notes
- [x] Add something to ~~developer~~ docs #27065 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
